core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -0,0 +1,275 @@
+  private static final Logger LOG = LoggerFactory.getLogger(UfsStatusCache.class);
+
+  private final ConcurrentHashMap<AlluxioURI, UfsStatus> mStatuses;
+  private final ConcurrentHashMap<AlluxioURI, Future<Collection<UfsStatus>>> mActivePrefetchJobs;
+  private final ConcurrentHashMap<AlluxioURI, Collection<UfsStatus>> mChildren;
+  private final ExecutorService mPrefetchExecutor;
+
+  /**
+   * Create a new instance of {@link UfsStatusCache}.
+   *
+   * @param prefetchExecutor the executor service used to prefetch statuses
+   */
+  public UfsStatusCache(@Nullable ExecutorService prefetchExecutor) {
+    mStatuses = new ConcurrentHashMap<>();
+    mChildren = new ConcurrentHashMap<>();
+    mActivePrefetchJobs = new ConcurrentHashMap<>();
+    mPrefetchExecutor = prefetchExecutor;
+  }
+
+  /**
+   * Add a new status to the cache.
+   *
+   * The last component of the path in the {@link AlluxioURI} must match the result of
+   * {@link UfsStatus#getName()}. This method overrides any status currently cached for the same
+   * URI.
+   *
+   * @param path the Alluxio path to key on
+   * @param status the ufs status to store
+   * @throws IllegalArgumentException if the status name doesn't match the final URI path component
+   */
+  public void addStatus(AlluxioURI path, UfsStatus status) {
+    UfsStatus prev = mStatuses.putIfAbsent(path, status);
+    if (!path.getName().equals(status.getName())) {
+      throw new IllegalArgumentException(
+          String.format("path name %s does not match ufs status name %s",
+              path.getName(), status.getName()));
+    }
+  }
+
+  /**
+   * Add a parent-child mapping to the status cache.
+   *
+   * All child statuses added via this method will be available via {@link #getStatus(AlluxioURI)}.
+   *
+   * @param path the directory inode path which contains the children
+   * @param children the children of the {@code path}
+   * @throws IllegalArgumentException when {@code path} already exists or if any child already
+   *                                  exists
+   */
+  public void addChildren(AlluxioURI path, Collection<UfsStatus> children) {
+    UfsStatus status = mStatuses.get(path);
+    // If this path doesn't yet exist, we can't keep track of the parent-child relationship
+    // We can still add statuses to the cache regardless
+    if (status != null) {
+      mChildren.computeIfAbsent(path, ufsStatus -> new ConcurrentHashSet<>()).addAll(children);
+    }
+    children.forEach(child -> {
+      AlluxioURI childPath = path.joinUnsafe(child.getName());
+      addStatus(childPath, child);
+    });
+  }
+
+  /**
+   * Remove a status from the cache.
+   *
+   *  Any children added to this status will remain in the cache.
+   *
+   * @param path the path corresponding to the {@link UfsStatus} to remove
+   * @return the removed UfsStatus
+   */
+  public UfsStatus remove(AlluxioURI path) {
+    UfsStatus removed = mStatuses.remove(path);
+    if (removed == null) {
```

Comment:
Is the contract that if the status DNE, then the children list DNE for that path? It is unclear what the relationship is between the 2 maps.

Reply:
the child map is a 1:N mapping of alluxio directory to UFS children.

If we remove the parent status from the cache, then it's assumed that we won't need to access its parent-child relationship information either. The UfsStatus information for each individual child still resides in mStatuses after calling `remove`

Reply:
Another solution is to make the parent-child relationship removal independent of the removing a single status, but then status removal will need 2 calls, and I think that is more prone to errors. We want to remove those references otherwise the JVM won't be able to GC the children if they are removed individually.

Reply:
1. Yes
2. Yes. Do you see thread-safety issues?

Reply:
in this case, are we ok with dangling cache entries? for a path /a/b/c, if /a is removed , /a/b -> {/a/b/c} is still in the cache, so is /a/b, do you think we should do cascading removal?

Reply:
let's say /a is removed and added later, it will suddenly get many entries from the past

Reply:
@ZacBlanco Is there any relationship between the listings cache, and the status cache? Meaning, if the listing cache has the mapping `/parent/ -> [/parent/a, /parent/b, /parent/c]`, is the status cache guaranteed to have the status of `/parent/a`, `/parent/b`, and `/parent/c`?

Reply:
@yuzhu  We don't want to do a cascading removal because it would be removing statuses from the cache that are potentially used in a later part of the sync.

The implementation is designed for the inode sync stream so that the statuses can be re-used later without needing to do more UFS listings. Additionally, the way this is used is that the parent is **always** synced and removed _before_ a child status even begins to sync.

We remove only the information that was needed for the parent to sync, while leaving all information about children in-tact so they don't need to call to the UFS when they are synced. Then the children then add information about their children to the cache, while removing theirs once finished, etc...

@gpang When initially added, they are guaranteed to be in sync, i.e. if `/parent -> {/parent/a, /parent/b, /parent/c}` exists then `/parent/a, /parent/b, /parent/c` will all exist in the cache individually. However, it's possible they _could_ be removed before the parent listing.

In the current code, the parent is **always** synced and removed before the child even begins to sync because we don't add children to the processing queue until the parent has completed all of its syncing.

Reply:
Also, I do see the issue with children still existing even if the parent doesn't.

There are not cases currently where that could happen as far as I'm aware in the current usage, but I meant to push an update for this when I changed the internal representation of the maps. Guess I missed it.